components.md

```diff
@@ -0,0 +1,98 @@
+It can be called with zero or more arguments.
+
+## Actor
+The Actor is a component which (typically) instantiates one or more Driver class and/or does some kind of processing.
+Actions which the actor can take could be anything from measuring on a regular basis ("acting" as a sensor), commanding hardware devices through the use of the driver to do X or Y, processing some data and re-emitting results (e.g. a PID controller).
+The actor is part of the communication network, that means it is listening for messages and maybe sending messages by itself.
+
+Potentially a GUI could be attached here too.
+
+The Actor would send commands to a Driver, instructing it to `set`/`get`/`call` a given Parameter/Action by name, with a value if appropriate, and receive replies.
+
+```{note}
+We could achieve more complete separation and make the Actor purely about processing _inputs_, doing some _operation_ and generating _outputs_.
+This could be either stateless (temperature conversion) or stateful (PID control).
+Maybe a better name would be "Processor" then.
+This means that _all_ device-related stuff stays compartmentalized in the Driver, and an Actor would address that Driver over the network (to fetch the Parameter values it is interested in).
+
+Thoughts?
+```
+
+## Driver
+A Driver is a component that interfaces with some hardware in a manner we don't specify, and that has a specific API on the ECP side.
+This might contain or wrap a `pymeasure.Instrument` instance or something from another instrument library.
+This is the place where all instrument libraries (including pymeasure) wire their hardware interface classes into ECP.
+
+Concerning the ECP, we draw the abstraction boundary at the Driver -- the details on how this communicates with hardware (SCPI, dll, ...) should not be relevant for the protocol details.
+What we define is how the other ECP components interact with the Driver, how it determines and announces its capabilities, etc.
+
+A driver may contain/manage
+* An object that communicates with the connected hardware device/instrument, including managing its lifecycle (init, operations, shutdown)
+* The name of the connected instrument
+* A list of available Parameters (properties/attributes to get/set) and Actions(methods to call)
+* A list of Parameters to poll/publish regularly (and the interval for that)
+* `set`/`get`/`get_all`/`call` interfaces (for incoming commands to use to act on Parameters and Actions)
+* A cache of parameter values (to avoid unnecessary communication)
+* Maybe a task queue and/or concurrent access management/locking
+* Logging configuration
+
+```{note}
+We might want to add the notion of "Channels", especially for the multi-Director stuff
+```
+
+# Parameter
+A Parameter is a property (in the English, not the Pythonic sense) of the device represented by a Driver
+It has a name and can be read(`get`) or `set`, and recent values might be cached in the Driver.
```

Comment:
Caching is a difficult topic: yes it is faster, but might not be true anymore.
I like pymeasure's stance to query the hardware, as that is the point of truth.
I see, however, that it might be good for some applications to cache values.

Reply:
I agree, that's why I said "might". In my tool I made good experience with it.
When the networks grow, we will have less control over the precise timing. If, let's say, 10 Parameters of an instrument are relevant/interesting, we might have Actors querying those every 500ms, a Observer for storage every 1000ms, a Director with a GUI every frame/16ms. That's a lot of requests and communication to re-fetch quite fresh data, which is what I want to avoid.

Another thing is that I had some instruments (registerbased) that could fetch the data for _all_ Parameters in _one_ request (very efficient!). I had a `get_all` method that made use of that for that instrument type, and updated all the cache values with that data, and the other participants (plot, PID, GUI) could then fetch the cached values really quickly.

We could foresee a configurable `cache_timeout` per Driver that controls how outdated data might become.

Reply:
Regular readout and publishing values as they become available can reduce device connection calls.
Nevertheless cached values are good:
I just see the case of a triggered device (oscilloscope), where you might want to know the last results without waiting for the next trigger.

Reply:
To be clear: I'm not against caching per se, but want to be careful.
And we have to weigh, whether the complexity and calls (calculate time difference...) are worth it.

Reply:
> I just see the case of a triggered device (oscilloscope), where you might want to know the last results without waiting for the next trigger.

my `get` method had a `from_cache=True` argument, so you could always force a fresh fetch if you need it.

Indeed, it needs weighing. I'll leave it in and we can refine later.